File: python/runkite_runner/store.py
"""Store Dual Mode for the Python runner.

Mirrors the checkpoint dual-mode pattern in checkpoint.py, applied to the
control plane's unified key-value store instead of checkpoints:

- direct mode (POSTGRES_DSN set): queries the control plane's own
  `store_items` table straight over psycopg -- same schema, same \\x1F
  namespace encoding as internal/state/postgres/postgres.go. Zero HTTP hop.
  Tenant comes from RunAssignment.tenant_id via tenant_ctx (bound per job
  in worker.py); absent/legacy assignments fall back to "default".
- proxy mode (no POSTGRES_DSN): calls the control plane's /internal/store/*
  HTTP API over httpx, sending X-Runkite-Tenant-Id from the same binding.
  Works against any backend (SQLite, Postgres).

Both modes read/write the exact same rows, so a value written by a Go
client via HTTP is immediately visible to a Python graph via direct mode,
and vice versa -- one store, not two competing systems.

`BaseStore` only requires `batch`/`abatch` to be implemented; every other
method (get/put/delete/search/list_namespaces, sync+async) is provided by
langgraph's base class in terms of these two.

Connection pooling (runner-side concurrency): direct mode used to hold
ONE shared psycopg connection serialized by an asyncio.Lock -- correct
under concurrent runs (the lock prevents interleaved writes on one
connection) but not actually parallel, since every store op from every
concurrent job queued behind that same lock. Now backed by a
psycopg_pool.AsyncConnectionPool sized to the runner's configured
concurrency (see worker.py's --concurrency / RunkiteStore's pool_size),
so concurrent jobs' store ops run on genuinely separate connections.

TTL support (real gap found via a live agent that called
store.aput(..., ttl=...): "TTL is not supported by RunkiteStore" --
deepagents / LangGraph BaseStore's own guard rejected the call before
it even reached batch/abatch, because this class never declared
supports_ttl). PutOp.ttl/GetOp.refresh_ttl/SearchOp.refresh_ttl
arrive already resolved by BaseStore (NOT_PROVIDED defaults applied via
self.ttl_config before construction) -- see _direct_one/_proxy_one's
PutOp/GetOp/SearchOp branches, which pass them straight through to the
control plane's store_items ttl_minutes/expires_at columns
(internal/state/*/*.go's PutItem/GetItem/SearchItems).
"""

from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import Iterable
from datetime import datetime, timedelta, timezone
from typing import Any

import httpx
from langgraph.store.base import (
    BaseStore,
    GetOp,
    Item,
    ListNamespacesOp,
    Op,
    PutOp,
    Result,
    SearchItem,
    SearchOp,
)

from . import pg_pool
from .tenant_ctx import current_tenant, tenant_headers
from .tls_utils import httpx_tls_kwargs

logger = logging.getLogger("runkite.store")

_NS_DELIM = "\x1f"


def _ns_to_string(ns: tuple[str, ...]) -> str:
    return _NS_DELIM + _NS_DELIM.join(ns) + _NS_DELIM


def _string_to_ns(s: str) -> tuple[str, ...]:
    trimmed = s.strip(_NS_DELIM)
    return tuple(trimmed.split(_NS_DELIM)) if trimmed else ()


def _ns_prefix_pattern(prefix: tuple[str, ...]) -> str:
    if not prefix:
        return "%"
    return _NS_DELIM + _NS_DELIM.join(prefix) + _NS_DELIM + "%"


def _timedelta_minutes(minutes: float) -> timedelta:
    return timedelta(minutes=minutes)


def _parse_ts(value: Any) -> datetime:
    if isinstance(value, datetime):
        return value
    if isinstance(value, str) and value:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    return datetime.now(timezone.utc)


def _item_from_json(d: dict) -> Item:
    return Item(
        namespace=tuple(d["namespace"] or ()),
        key=d["key"],
        value=d.get("value") or {},
        created_at=_parse_ts(d.get("created_at")),
        updated_at=_parse_ts(d.get("updated_at")),
    )


def _search_item_from_json(d: dict) -> SearchItem:
    return SearchItem(
        namespace=tuple(d["namespace"] or ()),
        key=d["key"],
        value=d.get("value") or {},
        created_at=_parse_ts(d.get("created_at")),
        updated_at=_parse_ts(d.get("updated_at")),
    )


def _split_match_conditions(
    match_conditions: tuple | None,
) -> tuple[tuple[str, ...] | None, tuple[str, ...] | None]:
    """Extract prefix/suffix paths, matching the control plane's LIKE-based
    matching (no wildcard-segment support on either side of the dual mode)."""
    prefix = suffix = None
    for cond in match_conditions or ():
        if cond.match_type == "prefix":
            prefix = tuple(cond.path)
        elif cond.match_type == "suffix":
            suffix = tuple(cond.path)
    return prefix, suffix


class RunkiteStore(BaseStore):
    """LangGraph BaseStore backed by the Runkite control plane's store_items table."""

    supports_ttl = True

    def __init__(
        self,
        *,
        postgres_dsn: str | None = None,
        http_base_url: str | None = None,
        runner_token: str | None = None,
        ttl_config: dict | None = None,
        pool_size: int = 4,
    ):
        if not postgres_dsn and not http_base_url:
            raise ValueError("RunkiteStore requires postgres_dsn or http_base_url")
        # BaseStore reads self.ttl_config to resolve NOT_PROVIDED ttl/
        # refresh_ttl defaults before PutOp/GetOp/SearchOp are even
        # constructed -- see this module's docstring.
        self.ttl_config = ttl_config
        self._dsn = postgres_dsn
        self._base_url = http_base_url.rstrip("/") if http_base_url else None
        self._headers: dict[str, str] = {}
        if runner_token:
            self._headers["X-Runner-Kind"] = "python-langgraph"
            self._headers["X-Runner-Token"] = runner_token
        self.mode = "direct" if postgres_dsn else "proxy"
        self._pool_size = pool_size
        self._pool = None  # AsyncConnectionPool, opened lazily on first direct-mode use
        # Event loop that owns the AsyncConnectionPool (and that sync
        # batch() must schedule onto). Set on first abatch/_get_pool.
        self._loop: asyncio.AbstractEventLoop | None = None
        # Guards only pool CREATION (a one-time, rare event), not every
        # store op -- unlike the old single-connection lock, this doesn't
        # serialize concurrent jobs' store ops against each other.
        self._pool_init_lock = asyncio.Lock()

    async def _get_pool(self):
        if self._pool is None:
            async with self._pool_init_lock:
                if self._pool is None:  # re-check: lost the race to open() below
                    pool = pg_pool.make(self._dsn, max_size=self._pool_size)
                    await pool.open()
                    self._pool = pool
                    self._loop = asyncio.get_running_loop()
        return self._pool

    async def _recreate_pool(self) -> None:
        """Drop a wedged pool (idle overnight / laptop sleep) and open a new one."""
        async with self._pool_init_lock:
            old = self._pool
            self._pool = None
            if old is not None:
                try:
                    await old.close()
                except Exception:
                    logger.exception("error closing wedged store pool")
        await self._get_pool()

    async def recover_if_wedged(self) -> None:
        """Cheap pre-job probe: if getconn hangs/fails, recreate once.

        Same overnight/laptop-sleep failure mode as _abatch_direct's
        PoolTimeout retry, but caught before the run pays a full 30s
        timeout on the first store op.
        """
        if self.mode != "direct" or self._pool is None:
            return
        from psycopg_pool import PoolTimeout

        try:
            async with self._pool.connection(timeout=5.0) as conn:
                await conn.execute("SELECT 1")
        except PoolTimeout:
            logger.warning("store pool timed out on health probe; recreating pool")
            await self._recreate_pool()
        except Exception:
            logger.warning("store pool health probe failed; recreating pool", exc_info=True)
            await self._recreate_pool()

    async def warm(self) -> None:
        """Bind to the current event loop and open the direct-mode pool.

        Must run once on the runner's main loop before jobs start. Opening
        the AsyncConnectionPool via asyncio.run() in a worker thread would
        leave the pool bound to a loop that dies when run() returns --
        later main-loop ops then hang on getconn. Proxy mode only records
        the loop (no pool). Also publishes the loop via runner_loop.bind
        so a RunkiteVectorStore constructed later in a job can schedule
        onto the same loop without its own warm() call.
        """
        from . import runner_loop

        self._loop = asyncio.get_running_loop()
        runner_loop.bind(self._loop)
        if self.mode == "direct":
            await self._get_pool()

    async def aclose(self) -> None:
        if self._pool is not None:
            await self._pool.close()
            self._pool = None

    # -- BaseStore abstract interface --------------------------------------

    def batch(self, ops: Iterable[Op]) -> list[Result]:
        # Sync BaseStore entrypoint. deepagents / LangGraph often call this
        # from asyncio.to_thread (a worker thread with NO running loop)
        # while the runner's AsyncConnectionPool was opened on the main
        # loop. asyncio.run(abatch) in that thread opens a *new* loop and
        # then hangs on pool.getconn (PoolTimeout after 30s) -- live SA
        # soak failure. Always schedule abatch onto the pool's owning
        # loop when it is known and still running.
        ops = list(ops)

        def _run_abatch() -> list[Result]:
            loop = self._loop
            if loop is not None and loop.is_running():
                return asyncio.run_coroutine_threadsafe(self.abatch(ops), loop).result(timeout=120)
            if self.mode == "direct":
                raise RuntimeError(
                    "RunkiteStore.batch() used before warm() on the runner event loop "
                    "(direct-mode pool must not be opened via asyncio.run in a worker thread)"
                )
            return asyncio.run(self.abatch(ops))

        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return _run_abatch()

        # Sync API invoked from inside a running loop: never block that
        # same loop waiting on itself -- hop to a worker thread that
        # run_coroutine_threadsafe's back onto the store loop.
        import concurrent.futures

        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
            return pool.submit(_run_abatch).result(timeout=120)

    async def abatch(self, ops: Iterable[Op]) -> list[Result]:
        if self._loop is None:
            self._loop = asyncio.get_running_loop()
        ops = list(ops)
        if self.mode == "direct":
            return await self._abatch_direct(ops)
        return await self._abatch_proxy(ops)

    # -- proxy mode: HTTP calls to the control plane -----------------------

    async def _abatch_proxy(self, ops: list[Op]) -> list[Result]:
        headers = {**self._headers, **tenant_headers()}
        async with httpx.AsyncClient(
            base_url=self._base_url, headers=headers, timeout=10.0, **httpx_tls_kwargs()
        ) as client:
            return [await self._proxy_one(client, op) for op in ops]

    async def _proxy_one(self, client: httpx.AsyncClient, op: Op) -> Result:
        # /internal/store/* (not the client-facing /store/*): a runner
        # authenticates with its runner token, not a client API key/JWT it
        # may not have. Same handlers on the Go side, different auth
        # boundary -- see internal/auth/auth.go.
        if isinstance(op, GetOp):
            resp = await client.get(
                "/internal/store/items",
                params={
                    "namespace": ",".join(op.namespace),
                    "key": op.key,
                    "refresh_ttl": "true" if op.refresh_ttl else "false",
                },
            )
            if resp.status_code == 404:
                return None
            resp.raise_for_status()
            return _item_from_json(resp.json())

        if isinstance(op, PutOp):
            if op.value is None:
                resp = await client.request(
                    "DELETE",
                    "/internal/store/items",
                    json={"namespace": list(op.namespace), "key": op.key},
                )
            else:
                resp = await client.put(
                    "/internal/store/items",
                    json={
                        "namespace": list(op.namespace),
                        "key": op.key,
                        "value": op.value,
                        "ttl_minutes": op.ttl,
                    },
                )
            resp.raise_for_status()
            return None

        if isinstance(op, SearchOp):
            resp = await client.post(
                "/internal/store/items/search",
                json={
                    "namespace_prefix": list(op.namespace_prefix),
                    "filter": op.filter or {},
                    "limit": op.limit,
                    "offset": op.offset,
                    "refresh_ttl": op.refresh_ttl,
                },
            )
            resp.raise_for_status()
            items = resp.json().get("items") or []
            return [_search_item_from_json(i) for i in items]

        if isinstance(op, ListNamespacesOp):
            prefix, suffix = _split_match_conditions(op.match_conditions)
            resp = await client.post(
                "/internal/store/namespaces",
                json={
                    "prefix": list(prefix) if prefix else None,
                    "suffix": list(suffix) if suffix else None,
                    "max_depth": op.max_depth,
                    "limit": op.limit,
                    "offset": op.offset,
                },
            )
            resp.raise_for_status()
            return [tuple(ns) for ns in (resp.json() or [])]

        raise TypeError(f"unsupported store op: {type(op)!r}")

    # -- direct mode: psycopg straight to store_items ------------------------

    async def _abatch_direct(self, ops: list[Op]) -> list[Result]:
        from psycopg_pool import PoolTimeout

        try:
            return await self._abatch_direct_once(ops)
        except PoolTimeout:
            # After long idle (laptop sleep, Postgres restart) the pool's
            # workers can wedge and every getconn hits the 30s timeout.
            # A closed pool cannot be reopen()'d -- rebuild and retry once.
            logger.warning("store pool timed out getting a connection; recreating pool")
            await self._recreate_pool()
            return await self._abatch_direct_once(ops)

    async def _abatch_direct_once(self, ops: list[Op]) -> list[Result]:
        pool = await self._get_pool()
        async with pool.connection() as conn:
            results: list[Result] = []
            for op in ops:
                results.append(await self._direct_one(conn, op))
            return results

    async def _direct_one(self, conn, op: Op) -> Result:
        if isinstance(op, GetOp):
            ns = _ns_to_string(op.namespace)
            async with conn.cursor() as cur:
                # Mirrors internal/state/postgres/postgres.go's GetItem:
                # an expired item (expires_at in the past) reads as
                # absent, and a refreshing read extends expires_at from
                # now using the item's own stored ttl_minutes.
                await cur.execute(
                    "SELECT namespace, key, value, created_at, updated_at, ttl_minutes "
                    "FROM store_items WHERE tenant_id = %s AND namespace = %s AND key = %s "
                    "AND (expires_at IS NULL OR expires_at > NOW())",
                    (current_tenant(), ns, op.key),
                )
                row = await cur.fetchone()
                if row is None:
                    return None
                ttl_minutes = row[5]
                if op.refresh_ttl and ttl_minutes is not None:
                    new_expiry = datetime.now(timezone.utc) + _timedelta_minutes(ttl_minutes)
                    await cur.execute(
                        "UPDATE store_items SET expires_at = %s WHERE tenant_id = %s AND namespace = %s AND key = %s",
                        (new_expiry, current_tenant(), ns, op.key),
                    )
                return _item_from_row(row[:5])

        if isinstance(op, PutOp):
            ns = _ns_to_string(op.namespace)
            # Computed in Python (not SQL interval arithmetic) to match
            # storeItemExpiresAt's approach on the Go side exactly, and
            # to avoid psycopg parameter-type ambiguity from reusing a
            # possibly-None placeholder in both a NULL check and interval
            # math.
            expires_at = None if op.ttl is None else datetime.now(timezone.utc) + _timedelta_minutes(op.ttl)
            async with conn.cursor() as cur:
                if op.value is None:
                    await cur.execute(
                        "DELETE FROM store_items WHERE tenant_id = %s AND namespace = %s AND key = %s",
                        (current_tenant(), ns, op.key),
                    )
                else:
                    await cur.execute(
                        """
                        INSERT INTO store_items (tenant_id, namespace, key, value, created_at, updated_at, ttl_minutes, expires_at)
                        VALUES (%s, %s, %s, %s, NOW(), NOW(), %s, %s)
                        ON CONFLICT (tenant_id, namespace, key)
                        DO UPDATE SET value = EXCLUDED.value, updated_at = NOW(),
                            ttl_minutes = EXCLUDED.ttl_minutes, expires_at = EXCLUDED.expires_at
                        """,
                        (current_tenant(), ns, op.key, json.dumps(op.value), op.ttl, expires_at),
                    )
            return None

        if isinstance(op, SearchOp):
            where = ["tenant_id = %s", "namespace LIKE %s", "(expires_at IS NULL OR expires_at > NOW())"]
            args: list[Any] = [current_tenant(), _ns_prefix_pattern(op.namespace_prefix)]
            for k, v in (op.filter or {}).items():
                where.append("value->>%s = %s")
                args.append(k)
                args.append(v if isinstance(v, str) else json.dumps(v))
            limit = op.limit or 10
            query = (
                "SELECT namespace, key, value, created_at, updated_at, ttl_minutes FROM store_items "
                f"WHERE {' AND '.join(where)} ORDER BY updated_at DESC LIMIT %s OFFSET %s"
            )
            args.extend([limit, op.offset])
            async with conn.cursor() as cur:
                await cur.execute(query, args)
                rows = await cur.fetchall()
                items = [_search_item_from_row(r[:5]) for r in rows]
                if op.refresh_ttl:
                    now = datetime.now(timezone.utc)
                    for r in rows:
                        ttl_minutes = r[5]
                        if ttl_minutes is not None:
                            await cur.execute(
                                "UPDATE store_items SET expires_at = %s WHERE tenant_id = %s AND namespace = %s AND key = %s",
                                (now + _timedelta_minutes(ttl_minutes), current_tenant(), r[0], r[1]),
                            )
                return items

        if isinstance(op, ListNamespacesOp):
            prefix, _suffix = _split_match_conditions(op.match_conditions)
            where = ["tenant_id = %s"]
            args = [current_tenant()]
            if prefix:
                where.append("namespace LIKE %s")
                args.append(_ns_prefix_pattern(prefix))
            query = "SELECT DISTINCT namespace FROM store_items WHERE " + " AND ".join(where)
            limit = op.limit or 100
            query += " ORDER BY namespace LIMIT %s OFFSET %s"
            args.extend([limit, op.offset])
            async with conn.cursor() as cur:
                await cur.execute(query, args)
                rows = await cur.fetchall()
                return [_string_to_ns(r[0]) for r in rows]

        raise TypeError(f"unsupported store op: {type(op)!r}")


def _item_from_row(row) -> Item:
    ns_str, key, value, created_at, updated_at = row
    return Item(
        namespace=_string_to_ns(ns_str),
        key=key,
        value=value if isinstance(value, dict) else json.loads(value or "{}"),
        created_at=_parse_ts(created_at),
        updated_at=_parse_ts(updated_at),
    )


def _search_item_from_row(row) -> SearchItem:
    item = _item_from_row(row)
    return SearchItem(
        namespace=item.namespace,
        key=item.key,
        value=item.value,
        created_at=item.created_at,
        updated_at=item.updated_at,
    )

File: typescript/runkite-runner/src/store.ts
/**
 * Store Dual Mode for the TypeScript runner. Direct TypeScript mirror of
 * the Python runner's store.py -- see that file's docstring for the full
 * rationale; repeated briefly here:
 *
 * - direct mode (POSTGRES_DSN set): queries the control plane's own
 *   store_items table straight over `pg` -- same schema, same \x1F
 *   namespace encoding as internal/state/postgres/postgres.go. Zero HTTP
 *   hop. Always operates in the "default" tenant (see TENANT_ID below) --
 *   same documented Direct Mode Trust Model trade-off as the Python
 *   runner: direct mode bypasses control-plane tenant scoping entirely.
 * - proxy mode (no POSTGRES_DSN): calls the control plane's /internal/
 *   store/* HTTP API. Works against any backend (SQLite, Postgres).
 *
 * Both modes read/write the exact same rows as the Go control plane and
 * the Python runner -- one store, not three competing systems.
 *
 * BaseStore only requires `batch` to be implemented; every other method
 * (get/put/delete/search/listNamespaces) is provided by langgraph's base
 * class in terms of it (see node_modules/@langchain/langgraph-checkpoint's
 * store/base.d.ts).
 */
import { BaseStore } from "@langchain/langgraph-checkpoint";
import type {
  GetOperation,
  Item,
  ListNamespacesOperation,
  Operation,
  OperationResults,
  PutOperation,
  SearchItem,
  SearchOperation,
} from "@langchain/langgraph-checkpoint";
import pg from "pg";
import { logger } from "./logger.js";
import { httpDispatcher, type FetchInit } from "./tls.js";

export const NS_DELIM = "\x1f";

/** Exported for direct unit testing of the round-trip/boundary-safety
 * properties -- same \x1F-delimited encoding as internal/state/postgres's
 * Go implementation and the Python runner's store.py, so all three must
 * agree on every edge case (empty namespace, segments containing "/",
 * prefix-vs-sibling boundary matching). */
export function nsToString(ns: readonly string[]): string {
  return NS_DELIM + ns.join(NS_DELIM) + NS_DELIM;
}

export function stringToNs(s: string): string[] {
  const trimmed = s.replace(new RegExp(`^${NS_DELIM}+|${NS_DELIM}+$`, "g"), "");
  return trimmed === "" ? [] : trimmed.split(NS_DELIM);
}

export function nsPrefixPattern(prefix: readonly string[]): string {
  if (prefix.length === 0) return "%";
  return NS_DELIM + prefix.join(NS_DELIM) + NS_DELIM + "%";
}

function parseTs(value: unknown): Date {
  if (value instanceof Date) return value;
  if (typeof value === "string" && value) return new Date(value);
  return new Date();
}

function itemFromRow(row: {
  namespace: string;
  key: string;
  value: unknown;
  created_at: unknown;
  updated_at: unknown;
}): Item {
  return {
    namespace: stringToNs(row.namespace),
    key: row.key,
    value: (typeof row.value === "string" ? JSON.parse(row.value || "{}") : row.value) ?? {},
    createdAt: parseTs(row.created_at),
    updatedAt: parseTs(row.updated_at),
  };
}

function isGetOp(op: Operation): op is GetOperation {
  return "namespace" in op && !("value" in op);
}
function isPutOp(op: Operation): op is PutOperation {
  return "namespace" in op && "value" in op;
}
function isSearchOp(op: Operation): op is SearchOperation {
  return "namespacePrefix" in op;
}
function isListNamespacesOp(op: Operation): op is ListNamespacesOperation {
  return !("namespace" in op) && !("namespacePrefix" in op);
}

export type StoreMode = "direct" | "proxy";

export class RunkiteStore extends BaseStore {
  mode: StoreMode;
  private pool: pg.Pool | null = null;
  private baseUrl: string | null = null;
  private headers: Record<string, string> = {};
  // Constructed once (not per-call): mirrors tls.ts's own doc comment --
  // undefined when RUNKITE_TLS_CA_FILE is unset, meaning "no dispatcher
  // option," which is fetch's own default behavior (plain http:// or
  // https:// with a publicly-trusted cert).
  private readonly dispatcher = httpDispatcher();

  // Direct mode has no per-request tenant identity to work with (it's a
  // raw DB connection, not an authenticated HTTP call) -- see the module
  // docstring's Direct Mode Trust Model note. Must match
  // internal/tenant.DefaultTenant on the Go side exactly.
  private static readonly TENANT_ID = "default";

  constructor(opts: { postgresDsn?: string; httpBaseUrl?: string; runnerToken?: string; poolSize?: number }) {
    super();
    if (!opts.postgresDsn && !opts.httpBaseUrl) {
      throw new Error("RunkiteStore requires postgresDsn or httpBaseUrl");
    }
    this.mode = opts.postgresDsn ? "direct" : "proxy";
    if (opts.postgresDsn) {
      // poolSize mirrors the Python runner's pool_size=concurrency --
      // see checkpoint.ts's CheckpointerManager.start() for the full
      // rationale. Undefined leaves node-postgres's own default (10).
      // See checkpoint.ts for idle/connect timeout rationale.
      this.pool = new pg.Pool({
        connectionString: opts.postgresDsn,
        idleTimeoutMillis: 60_000,
        connectionTimeoutMillis: 10_000,
        ...(opts.poolSize ? { max: opts.poolSize } : {}),
      });
      this.pool.on("error", (err) => {
        // Idle clients can error after server-side close; log and let
        // the pool drop them rather than crashing the process.
        logger.error(`store pool idle client error: ${err.message}`);
      });
    }
    if (opts.httpBaseUrl) {
      this.baseUrl = opts.httpBaseUrl.replace(/\/+$/, "");
    }
    if (opts.runnerToken) {
      this.headers["X-Runner-Kind"] = "typescript-langgraphjs";
      this.headers["X-Runner-Token"] = opts.runnerToken;
    }
  }

  async close(): Promise<void> {
    if (this.pool) await this.pool.end();
  }

  async batch<Op extends Operation[]>(operations: Op): Promise<OperationResults<Op>> {
    const results = await Promise.all(
      operations.map((op) => (this.mode === "direct" ? this.directOne(op) : this.proxyOne(op))),
    );
    return results as OperationResults<Op>;
  }

  // -- direct mode: pg straight to store_items -----------------------------

  private async directOne(op: Operation): Promise<unknown> {
    const pool = this.pool!;
    if (isGetOp(op)) {
      // Hide expired rows the same way Python/Go stores do.
      const res = await pool.query(
        `SELECT namespace, key, value, created_at, updated_at FROM store_items
         WHERE tenant_id = $1 AND namespace = $2 AND key = $3
           AND (expires_at IS NULL OR expires_at > NOW())`,
        [RunkiteStore.TENANT_ID, nsToString(op.namespace), op.key],
      );
      return res.rows[0] ? itemFromRow(res.rows[0]) : null;
    }

    if (isPutOp(op)) {
      const ns = nsToString(op.namespace);
      if (op.value === null) {
        await pool.query("DELETE FROM store_items WHERE tenant_id = $1 AND namespace = $2 AND key = $3", [
          RunkiteStore.TENANT_ID,
          ns,
          op.key,
        ]);
      } else {
        // LangGraph JS PutOperation typings omit ttl today; accept it when
        // present (parity with Python runner / Go store_items columns).
        const ttlMinutes = (op as PutOperation & { ttl?: number }).ttl;
        const expiresAt =
          ttlMinutes == null ? null : new Date(Date.now() + ttlMinutes * 60_000);
        await pool.query(
          `INSERT INTO store_items (tenant_id, namespace, key, value, created_at, updated_at, ttl_minutes, expires_at)
           VALUES ($1, $2, $3, $4, NOW(), NOW(), $5, $6)
           ON CONFLICT (tenant_id, namespace, key)
           DO UPDATE SET value = EXCLUDED.value, updated_at = NOW(),
             ttl_minutes = EXCLUDED.ttl_minutes, expires_at = EXCLUDED.expires_at`,
          [RunkiteStore.TENANT_ID, ns, op.key, JSON.stringify(op.value), ttlMinutes ?? null, expiresAt],
        );
      }
      return undefined;
    }

    if (isSearchOp(op)) {
      const where = ["tenant_id = $1", "namespace LIKE $2", "(expires_at IS NULL OR expires_at > NOW())"];
      const args: unknown[] = [RunkiteStore.TENANT_ID, nsPrefixPattern(op.namespacePrefix)];
      let argN = 3;
      for (const [k, v] of Object.entries(op.filter ?? {})) {
        where.push(`value->>$${argN} = $${argN + 1}`);
        args.push(k, typeof v === "string" ? v : JSON.stringify(v));
        argN += 2;
      }
      const limit = op.limit ?? 10;
      const query =
        `SELECT namespace, key, value, created_at, updated_at FROM store_items ` +
        `WHERE ${where.join(" AND ")} ORDER BY updated_at DESC LIMIT $${argN} OFFSET $${argN + 1}`;
      args.push(limit, op.offset ?? 0);
      const res = await pool.query(query, args);
      return res.rows.map((r) => itemFromRow(r) as SearchItem);
    }

    if (isListNamespacesOp(op)) {
      const prefix = op.matchConditions?.find((c) => c.matchType === "prefix")?.path as string[] | undefined;
      const where = ["tenant_id = $1"];
      const args: unknown[] = [RunkiteStore.TENANT_ID];
      let argN = 2;
      if (prefix) {
        where.push(`namespace LIKE $${argN}`);
        args.push(nsPrefixPattern(prefix));
        argN++;
      }
      const query = `SELECT DISTINCT namespace FROM store_items WHERE ${where.join(" AND ")} ORDER BY namespace LIMIT $${argN} OFFSET $${argN + 1}`;
      args.push(op.limit ?? 100, op.offset ?? 0);
      const res = await pool.query(query, args);
      return res.rows.map((r) => stringToNs(r.namespace));
    }

    throw new TypeError(`unsupported store op: ${JSON.stringify(op)}`);
  }

  // -- proxy mode: HTTP calls to the control plane -------------------------

  private async proxyOne(op: Operation): Promise<unknown> {
    // /internal/store/* (not the client-facing /store/*): a runner
    // authenticates with its runner token, not a client API key/JWT it
    // may not have. Same handlers on the Go side, different auth
    // boundary -- see internal/auth/auth.go.
    if (isGetOp(op)) {
      const url = `${this.baseUrl}/internal/store/items?namespace=${encodeURIComponent(op.namespace.join(","))}&key=${encodeURIComponent(op.key)}`;
      const opts: FetchInit = { headers: this.headers, dispatcher: this.dispatcher };
      const resp = await fetch(url, opts);
      if (resp.status === 404) return null;
      if (!resp.ok) throw new Error(`GET store item failed: ${resp.status} ${await resp.text()}`);
      return itemFromJson(await resp.json());
    }

    if (isPutOp(op)) {
      if (op.value === null) {
        const opts: FetchInit = {
          method: "DELETE",
          headers: { ...this.headers, "Content-Type": "application/json" },
          body: JSON.stringify({ namespace: op.namespace, key: op.key }),
          dispatcher: this.dispatcher,
        };
        const resp = await fetch(`${this.baseUrl}/internal/store/items`, opts);
        if (!resp.ok) throw new Error(`DELETE store item failed: ${resp.status} ${await resp.text()}`);
      } else {
        // LangGraph JS PutOperation typings omit ttl; forward when present
        // (matches Python proxy + Go StorePutRequest.ttl_minutes).
        const ttlMinutes = (op as PutOperation & { ttl?: number }).ttl;
        const body: Record<string, unknown> = {
          namespace: op.namespace,
          key: op.key,
          value: op.value,
        };
        if (ttlMinutes != null) body.ttl_minutes = ttlMinutes;
        const opts: FetchInit = {
          method: "PUT",
          headers: { ...this.headers, "Content-Type": "application/json" },
          body: JSON.stringify(body),
          dispatcher: this.dispatcher,
        };
        const resp = await fetch(`${this.baseUrl}/internal/store/items`, opts);
        if (!resp.ok) throw new Error(`PUT store item failed: ${resp.status} ${await resp.text()}`);
      }
      return undefined;
    }

    if (isSearchOp(op)) {
      const opts: FetchInit = {
        method: "POST",
        headers: { ...this.headers, "Content-Type": "application/json" },
        body: JSON.stringify({
          namespace_prefix: op.namespacePrefix,
          filter: op.filter ?? {},
          limit: op.limit,
          offset: op.offset,
        }),
        dispatcher: this.dispatcher,
      };
      const resp = await fetch(`${this.baseUrl}/internal/store/items/search`, opts);
      if (!resp.ok) throw new Error(`search store items failed: ${resp.status} ${await resp.text()}`);
      const body = (await resp.json()) as { items?: unknown[] };
      return (body.items ?? []).map((i) => itemFromJson(i) as SearchItem);
    }

    if (isListNamespacesOp(op)) {
      const prefix = op.matchConditions?.find((c) => c.matchType === "prefix")?.path as string[] | undefined;
      const suffix = op.matchConditions?.find((c) => c.matchType === "suffix")?.path as string[] | undefined;
      const opts: FetchInit = {
        method: "POST",
        headers: { ...this.headers, "Content-Type": "application/json" },
        body: JSON.stringify({
          prefix: prefix ?? null,
          suffix: suffix ?? null,
          max_depth: op.maxDepth ?? null,
          limit: op.limit,
          offset: op.offset,
        }),
        dispatcher: this.dispatcher,
      };
      const resp = await fetch(`${this.baseUrl}/internal/store/namespaces`, opts);
      if (!resp.ok) throw new Error(`list namespaces failed: ${resp.status} ${await resp.text()}`);
      const body = (await resp.json()) as string[][] | null;
      return body ?? [];
    }

    throw new TypeError(`unsupported store op: ${JSON.stringify(op)}`);
  }
}

function itemFromJson(d: any): Item {
  return {
    namespace: d.namespace ?? [],
    key: d.key,
    value: d.value ?? {},
    createdAt: parseTs(d.created_at),
    updatedAt: parseTs(d.updated_at),
  };
}
